### PR TITLE
Explicit modifiers ("hyper key")

### PR DESCRIPTION
**Pull request by okke-formsma**
Also see #361. 

This PR makes LS(LEFT_CONTROL) work as if shift and control were both
pressed explicitly. Previously, the left shift would have been released
as soon as another key was pressed. The implicit behavior is useful in
case of LS(NUMBER_1) when rolling over to other keys.

There are also a few refactor commits included, which clean up related code.

### CODE REVIEW

**Comment by mrzealot** [comment]
Did a quick test with my `&mt LC(LS(LALT)) RET` key, works like a charm now. Would shop here again.

**Comment by okke-formsma** [comment]
rebased onto 2.4.0

File: app/include/dt-bindings/zmk/modifiers.h [hunk 1]
```diff
@@ -25,4 +25,8 @@
 #define RC(keycode) APPLY_MODS(MOD_RCTL, keycode)
 #define RS(keycode) APPLY_MODS(MOD_RSFT, keycode)
 #define RA(keycode) APPLY_MODS(MOD_RALT, keycode)
-#define RG(keycode) APPLY_MODS(MOD_RGUI, keycode)
\ No newline at end of file
+#define RG(keycode) APPLY_MODS(MOD_RGUI, keycode)
+
+#define IS_MOD(usage_page, keycode)                                                                \
+    (usage_page == HID_USAGE_KEY && keycode >= HID_USAGE_KEY_KEYBOARD_LEFTCONTROL &&               \
+     keycode <= HID_USAGE_KEY_KEYBOARD_RIGHT_GUI)
```
Comment (innovaker) [inline app/include/dt-bindings/zmk/modifiers.h:32]: Should this be a macro or a function?  @joelspadin, any thoughts please?

File: app/src/hid.c [hunk 2]
```diff
@@ -49,6 +49,32 @@ int zmk_hid_unregister_mod(zmk_mod modifier) {
     return 0;
 }
 
+int zmk_hid_register_mods(zmk_mod_flags modifiers) {
+    int err;
+    for (zmk_mod i = 0; i < 8; i++) {
```
Comment (petejohanson) [inline app/src/hid.c:54]: `sizeof(zmk_mod_flags_t)`?

File: app/src/hid.c [hunk 3]
```diff
@@ -49,6 +49,32 @@ int zmk_hid_unregister_mod(zmk_mod modifier) {
...
+    }
+    return 0;
+}
+
+int zmk_hid_unregister_mods(zmk_mod_flags modifiers) {
+    int err;
+    for (zmk_mod i = 0; i < 8; i++) {
+        if (modifiers & (1 << i)) {
+            err = zmk_hid_unregister_mod(i);
+            if (err) {
```
Comment (petejohanson) [inline app/src/hid.c:70]: Do we really want to bail on the loop if some entry errors?

**Review by petejohanson — COMMENTED** [review]
A couple thoughts.

File: app/include/dt-bindings/zmk/modifiers.h (hunk 1 above)
Comment (joelspadin) [reply app/include/dt-bindings/zmk/modifiers.h:32]: I don't see anything in here that would require it to be a macro, so my vote is function. I can almost guarantee you that the compiler will inline this if you make it a function, and the resulting binary will be very close to if not identical to with a macro.

File: app/src/hid.c (hunk 3 above)
Comment (okke-formsma) [reply app/src/hid.c:70]: Nah. There's no way to recover from an error either, so I'll remove the error handling.

File: app/include/dt-bindings/zmk/modifiers.h (hunk 1 above)
Comment (okke-formsma) [reply app/include/dt-bindings/zmk/modifiers.h:32]: would you define that function in the header file, and if not, where should we put it? I would like to keep this close together with the `SELECT_MODS` macro and friends.

Comment (joelspadin) [reply app/include/dt-bindings/zmk/modifiers.h:32]: Two ways you can do this:

1. Define it in the header as an inline function. (Needs to be inline so the linker won't complain when it sees that every C file that includes modifiers.h has its own copy of the function.)
2. Declare it in the header and define it in some other C file.

Comment (petejohanson) [reply app/include/dt-bindings/zmk/modifiers.h:32]: So, the challenge here, is this header here *really* should only be macros, and those for keymaps. Either way this code shouldn't really be here. Perhaps inline function in `app/include/zmk/keys.h`?

**Review by petejohanson — COMMENTED** [review]
On a whole, this looks great. Just one comment on the macro vs. inline function discussion.

**Review by petejohanson — APPROVED** [review]

**Comment by petejohanson** [comment]
@okke-formsma can you rebase to fix merge conflicts?

**Comment by okke-formsma** [comment]
@petejohanson done

**Pull request merged**